cmd/snap-seccomp/main.go

```diff
@@ -789,28 +795,132 @@ func complainAction() seccomp.ScmpAction {
 	return seccomp.ActAllow
 }
 
+var osCreateTemp = os.CreateTemp
+
+func exportBPF(fout *os.File, filter *seccomp.ScmpFilter) (bpfLen int64, err error) {
+	errPrefixFmt := "cannot export bpf filter: %w"
+
+	oldPos, err := fout.Seek(0, os.SEEK_CUR)
+	if err != nil {
+		return 0, fmt.Errorf(errPrefixFmt, err)
+	}
+	if err := filter.ExportBPF(fout); err != nil {
+		return 0, fmt.Errorf(errPrefixFmt, err)
+	}
+	nowPos, err := fout.Seek(0, os.SEEK_CUR)
+	if err != nil {
+		return 0, fmt.Errorf(errPrefixFmt, err)
+	}
+
+	return nowPos - oldPos, nil
+}
+
+// New .bin2 seccomp files are composed by the following header, and potentially one
+// allow filter and/or one deny filter (if lenAllowFilter and lenDenyFilter are greater
+// than 0 respectively). When more than one filter is loaded, the kernel applies
+// the most restrictive action, thus any explicit deny will take precedence.
+// This struct needs to be in sync with seccomp-support.c
+type scSeccompFileHeader struct {
+	header  [2]byte
+	version byte
+	// flags
+	unrestricted byte
+	// unused
+	padding [4]byte
+	// location of allow/deny, all offsets/len in bytes
+	lenAllowFilter uint32
+	lenDenyFilter  uint32
+	// reserved for future use
+	reserved2 [112]byte
+}
+
+func writeUnrestrictedFilter(outFile string) error {
+	hdr := scSeccompFileHeader{
+		header:  [2]byte{'S', 'C'},
+		version: 0x1,
+		// tell snap-confine
+		unrestricted: 0x1,
+	}
+	fout, err := osutil.NewAtomicFile(outFile, 0644, 0, osutil.NoChown, osutil.NoChown)
+	if err != nil {
+		return err
+	}
+	defer fout.Cancel()
+
+	if err := binary.Write(fout, arch.Endian(), hdr); err != nil {
+		return err
+	}
+	return fout.Commit()
+}
+
+func writeSeccompFilter(outFile string, filterAllow, filterDeny *seccomp.ScmpFilter) error {
+	fout, err := osutil.NewAtomicFile(outFile, 0644, 0, osutil.NoChown, osutil.NoChown)
+	if err != nil {
+		return err
+	}
+	defer fout.Cancel()
+
+	// Write preliminary header because we don't know the sizes of the
+	// seccomp filters yet and the only way to know is to export to
+	// a file (until seccomp_export_bpf_mem() becomes available)
+	hdr := scSeccompFileHeader{
+		header:  [2]byte{'S', 'C'},
+		version: 0x1,
+	}
+	if err := binary.Write(fout, arch.Endian(), hdr); err != nil {
```

Comment:
Could we just fix the header and not overwrite it? Move the two export calls above, get the sizes into the header and we're done.

Reply:
Calls to export already write the filter to the file. Thus, writing the 'dummy' header before calling export is needed to reserve the space at the beginning of the file. We could then seek for the position of the filters len in the header and only overwrite the lengths or use staging files to get the length of the filter before writing them to its final location, but it will bring extra complexity without benefits from my point of view